integration-tests/features/steps/three_scale_preview.py

```diff
@@ -0,0 +1,167 @@
+def running_component_search_api(context):
+    """Check three scale preview service is available."""
+    assert context.is_3scale_preview_running(context)
+
+
+@when("I send NPM package manifest {manifest} to stack analysis through "
+      "3scale gateway {user_key} user_key")
+@when("I send NPM package manifest {manifest} to stack analysis {rate:d} times"
+      " in a minute through 3scale gateway {user_key} user_key")
+def npm_manifest_stack_analysis(context, manifest, version=3, user_key="without", rate=1):
+    """Send the NPM package manifest file to the stack analyses."""
+    endpoint = threescale_preview_endpoint_url(context, 'stack-analyses')
+    use_user_key = parse_token_clause(user_key)
+    send_manifest_to_stack_analyses(context, manifest, 'package.json',
+                                    endpoint, use_user_key, rate)
+
+
+@when("I call stack analysis {rate:d} times in a minute {user_key} user_key")
+@when("I wait for stack analysis to finish {user_key} user_key")
+def wait_for_stack_analysis_completion(context, user_key="without", rate=1):
+    """Try to wait for the stack analysis to be finished.
+
+    This step assumes that stack analysis has been started previously and
+    thus that the job ID is known
+
+    Current API implementation returns just three HTTP codes:
+    200 OK : analysis is already finished
+    202 Accepted: analysis is started or is in progress (or other state!)
+    403 UNAUTHORIZED : missing or improper user_key
+    """
+    timeout = context.stack_analysis_timeout  # in seconds
+    sleep_amount = 15  # we don't have to overload the API with too many calls
+    use_user_key = parse_token_clause(user_key)
+
+    id = context.response.json().get("id")
+    context.stack_analysis_id = id
+    url = urljoin(threescale_preview_endpoint_url(context, 'stack-analyses'), id)
+
+    for _ in range((timeout // sleep_amount) + rate):
+        if rate > 1:
+            sleep_amount = 0  # to get rate limit exceeded we have to overload API calls
+        if use_user_key:
+            context.response = requests.get(
+                url, params={'user_key': context.three_scale_preview_user_key})
+        else:
+            context.response = requests.get(url)
+        status_code = context.response.status_code
+        if status_code == 200:
+            json_resp = context.response.json()
+            if contains_alternate_node(json_resp) and rate <= 1:
+                break
+        # 429 (Rate Limit) code should be checked later
+        elif status_code == 429:
+            break
+        # 403 code should be checked later
+        elif status_code == 403:
+            break
+        elif status_code != 202:
+            raise Exception('Bad HTTP status code {c}'.format(c=status_code))
+        time.sleep(sleep_amount)
+    else:
+        raise Exception('Timeout waiting for the stack analysis results')
+
+
+@then("I should get {response_txt} text response")
```

Comment:
```suggestion
@then("I should get {response_txt} text response")
```

this function is common, so might not have 'rate limit' in docstring.